### PR TITLE
[VOL-83] Show running containers in `list` output

### PR DESCRIPTION
**Pull request by zubron**
Fixes [VOL-83](https://clusterhq.atlassian.net/browse/VOL-83).

This change adds support for showing running containers using a dvol volume in the output of `dvol list`. To enable this, I've added a new dependency on [go-dockerclient](https://github.com/fsouza/go-dockerclient). To support the disabling of docker integration, I've added a new parameter to `NewDvolAPI`. There might be a nicer way to do this so I'd appreciate feedback or ideas on this in the review :)

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -74,9 +77,16 @@ type DvolVolume struct {
 	Path string
 }
 
-func NewDvolAPI(basePath string) *DvolAPI {
+func NewDvolAPI(basePath string, disableDockerIntegration bool) *DvolAPI {
 	dl := datalayer.NewDataLayer(basePath)
-	return &DvolAPI{basePath, dl}
+	var cr containers.Runtime
```
Comment (brendan-donegan) [inline pkg/api/api.go]: can we call this containerRuntime?

File: pkg/cmd/list.go [hunk 2]
```diff
@@ -60,7 +60,12 @@ func listVolumes(cmd *cobra.Command, args []string, out io.Writer) error {
 		if err != nil {
 			return err
 		}
-		if _, err := fmt.Fprintf(writer, "%s%s\t%s\t\n", prefix, volume.Name, branch); err != nil {
+		c, err := dvol.ContainerRuntime.Related(volume.Name)
```
Comment (brendan-donegan) [inline pkg/cmd/list.go]: c is for? 'containers'?

File: pkg/cmd/list.go [hunk 3]
```diff
@@ -60,7 +60,12 @@ func listVolumes(cmd *cobra.Command, args []string, out io.Writer) error {
 		if err != nil {
 			return err
 		}
-		if _, err := fmt.Fprintf(writer, "%s%s\t%s\t\n", prefix, volume.Name, branch); err != nil {
+		containers, err := dvol.ContainerRuntime.Related(volume.Name)
```
Comment (robhaswell) [inline pkg/cmd/list.go]: I don't like reaching into APIs. I think `ContainerRuntime` should be private and this call should be `dvol.relatedContainers(volume.Name)`.

File: pkg/containers/docker.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+func isRelated(volume string, container *docker.Container) bool {
+	for _, mount := range container.Mounts {
+		if mount.Name == volume && mount.Driver == "dvol" {
+			return true
+		}
+	}
+	return false
+}
+
+func (runtime DockerRuntime) Related(volume string) ([]string, error) {
```
Comment (robhaswell) [inline pkg/containers/docker.go]: This method only returns related and _running_ containers, which is a side-effect of what we intend to do with them (restart them). I think that's OK with the current method name, although I would refer e.g. `RelatedRunning`. However I think at least a docstring mentioning this fact would be useful to future developers.

File: pkg/containers/docker.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	containers, _ := runtime.Client.ListContainers(docker.ListContainersOptions{})
```
Comment (robhaswell) [inline pkg/containers/docker.go]: This can probably return a meaningful error.

File: dvol_python/test_plugin.py [hunk 6]
```diff
@@ -83,7 +83,6 @@ def dvol_list_includes_memorydiskserver():
                 raise Exception("volume never showed up in result %s" % (result,))
         try_until(dvol_list_includes_memorydiskserver)
 
-    @skip_if_go_version
     def test_docker_run_dvol_container_show_up_in_list_output(self):
         container = "fancy"
         def cleanup():
```
Comment (robhaswell) [inline dvol_python/test_plugin.py]: This test and the one beneath it has potential bugs in the cleanup code. If the first run call raises, the second run call will not, meaning memorydiskserver is not removed. It has been fixed in the other tests, please could you update these ones as well?

**Comment by robhaswell** [comment]
This branch looks great - thanks. Please could you address the comments and also this discussion on DvolAPIOptions that we had in Slack? I think I would like to re-review after that refactoring.

Thanks!

File: dvol_python/test_plugin.py [hunk 7]
```diff
@@ -83,16 +83,21 @@ def dvol_list_includes_memorydiskserver():
...
+                ["docker", "volume", "rm", "memorydiskserver"],
+                [DVOL, "rm", "-f", "memorydiskserver"],
+            ]
+            for cmd in cmds:
+                try:
+                    run(cmd)
+                except:
+                    pass
+        cleanup()
+        self.addCleanup(cleanup)
```
Comment (robhaswell) [inline dvol_python/test_plugin.py]: You need to remove the manual `cleanup()` at the bottom of this file.

File: dvol_python/test_plugin.py [hunk 8]
```diff
@@ -105,18 +110,23 @@ def dvol_list_includes_container_name():
...
```
Comment (robhaswell) [inline dvol_python/test_plugin.py]: Again, remove the manual `cleanup()`.

**Comment by robhaswell** [comment]
Thanks @zubron, nice work on the refactoring. I don't have any more comments other than those which you addressed, so please merge once Travis is happy. I have looked at the errors from Travis and they seem spurious, so please investigate with whomever our buildmaster is.

**Pull request merged**